sklearn/semi_supervised/self_training.py

```diff
@@ -0,0 +1,239 @@
+
+    X_labeled_ : array, shape = (n_samples, n_features)
+        The labeled samples used for the final fit.
+
+    y_labeled_iter_ : array, shape = (n_samples)
+        The iteration in which each sample was labeled. When a sample has
+        iteration 0, the sample was labeled in the given dataset.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn import datasets
+    >>> from sklearn.semi_supervised import SelfTrainingClassifier
+    >>> from sklearn.svm import SVC
+    >>> svc = SVC(probability=True, gamma="auto")
+    >>> self_training_model = SelfTrainingClassifier(svc)
+    >>> iris = datasets.load_iris()
+    >>> rng = np.random.RandomState(42)
+    >>> random_unlabeled_points = rng.rand(len(iris.target)) < 0.3
+    >>> labels = np.copy(iris.target)
+    >>> labels[random_unlabeled_points] = -1
+    >>> self_training_model.fit(iris.data, labels)
+    ... # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    SelfTrainingClassifier(...)
+
+    References
+    ----------
+    David Yarowsky. 1995. Unsupervised word sense disambiguation rivaling
+    supervised methods. In Proceedings of the 33rd annual meeting on
+    Association for Computational Linguistics (ACL '95). Association for
+    Computational Linguistics, Stroudsburg, PA, USA, 189-196. DOI:
+    https://doi.org/10.3115/981658.981684
+    """
+    def __init__(self,
+                 base_estimator,
+                 threshold=0.75,
+                 max_iter=100):
+        self.base_estimator = base_estimator
+        self.threshold = threshold
+        self.max_iter = max_iter
+
+    def fit(self, X, y):
+        """
+        Fits this ``SelfTrainingClassifier`` to dataset, using the
+        base_estimator passed. The fitted base_estimator is stored as
+        ``self.base_estimator_``.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+            array representing the data
+
+        y : array-like, shape = (n_samples, 1)
+            array representing the labels
+
+        Returns
+        -------
+        self: returns an instance of self.
+        """
+        X, y = check_X_y(X, y)
+        _check_estimator(self.base_estimator)
+        self.base_estimator_ = clone(self.base_estimator)
+
+        if not 0 <= self.max_iter:
+            raise ValueError("max_iter must be >= 0")
+
+        if not 0 <= self.threshold < 1:
+            raise ValueError("threshold must be in [0,1)")
+
+        # Data usable for supervised training
+        self.X_labeled_ = X[safe_mask(X, np.where(y != -1))][0]
+        self.y_labeled_ = y[safe_mask(y, np.where(y != -1))][0]
+        self.y_labeled_iter_ = np.full_like(self.y_labeled_, 0)
+
+        # Unlabeled data
+        X_unlabeled = X[safe_mask(X, np.where(y == -1))][0]
+        y_unlabeled = y[safe_mask(y, np.where(y == -1))][0]
+
+        iter = 0
+        while len(self.X_labeled_) < len(X) and iter < self.max_iter:
```

Comment:
We should store the number of iterations to convergence as `self.n_iter_` for consistency with other iterative algorithms.

Should we raise a ConvergenceWarning if max_iter is reached without labelling all points?

One way to code this up is:

```py
for self.n_iter_ in range(1, self.max_iter + 1):
    ...
    if len(self.X_labeled_) == len(X):
        break
else:
    warnings.warn(ConvergenceWarning, ...)
```

Reply:
I think its reasonable for the classifier to not have labeled all the datapoints, since ideally, we want to enrich the dataset with correct data, and don't mind uncertain datapoints remaining unlabeled. I feel like raising a warning would give users an indication that something is wrong, and perhaps even prompt them to do things like lower the threshold until the warning is gone, even though I found that having some unlabeled datapoints at the end is quite common.

However, I'm not familiar enough with ConvergenceWarnings in sklearn to know when they are supposed to be thrown, so I'd leave this decision up to you.

Reply:
I had thought the same. Let's not have a ConvergenceWarning.